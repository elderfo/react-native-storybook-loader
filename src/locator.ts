import glob from 'glob';
import path from 'path';

import { formatPath, getRelativePath, stripExtension } from './paths';
import { Configuration } from './configuration';

export type LoaderDefinition = {
  outputFile: string;
  storyFiles: string[];
};

export const generateLoaderDefinition = async ({
  rootDirectory,
  outputFile,
  searchDir,
  pattern,
}: Configuration): Promise<LoaderDefinition> => {
  const fullOutputFile = path.resolve(rootDirectory, outputFile);
  const outputFileDir = path.dirname(fullOutputFile);

  const lookupPatterns: string[] = []
  searchDir.forEach(dir =>
    pattern.forEach(p =>
      lookupPatterns.push(path.resolve(rootDirectory, dir, p))
    )
  )
  
  const lookupFiles = lookupPatterns
    .map(f => formatPath(f))
    .reduce((acc: string[], p: string) => [...acc, ...glob.sync(p)], [])
    // Applying a format again to ensure paths are using '/'
    .map(file => formatPath(file));

  const uniqueFiles = Array.from(new Set(lookupFiles));

  return {
    outputFile: fullOutputFile,
    storyFiles: uniqueFiles
      .map(f => getRelativePath(f, outputFileDir))
      .map(f => stripExtension(f))
      .map(f => formatPath(f)),
  };
};
