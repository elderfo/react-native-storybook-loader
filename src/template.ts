import prettier from 'prettier';

import { LoaderDefinition } from './locator';
import logger from './logger';

const formatter = (
  files: string[],
  frms: (file: string) => string,
  separator: string
) => {
  const formatted = files.map(f => frms(f));
  return formatted.join(separator);
};

const defaultPrettierOptions: prettier.Options = {
  parser: 'babel',
};

export const generateTemplate = async (
  loader: LoaderDefinition
): Promise<string> => {
  const template = `// Auto-generated file created by react-native-storybook-loader
  // Do not edit.
  //
  // https://github.com/elderfo/react-native-storybook-loader.git
  
  function loadStories() {
  ${formatter(loader.storyFiles, file => `require('${file}');`, '\n')}
  }
  
  const stories = [
  ${formatter(loader.storyFiles, file => `'${file}'`, ',\n')}
  ];
  
  module.exports = {
    loadStories,
    stories,
  };
  `;

  return await makePrettier(template);
};

const makePrettier = async (template: string): Promise<string> => {
  try {
    const prettierConfigFile = await prettier.resolveConfigFile();

    if (prettierConfigFile === null) {
      logger.info(
        'Prettier configuration not detected, using default formatting.'
      );
      return prettier.format(template, defaultPrettierOptions);
    }

    logger.info(
      `Attempting to use prettier configuration detected at ${prettierConfigFile}`
    );

    const prettierConfig = await prettier.resolveConfig(prettierConfigFile);

    if (prettierConfig === null) {
      logger.warn(
        'Prettier configuration was not found, using default formatting.'
      );
      return prettier.format(template, defaultPrettierOptions);
    }

    // Since many prettier configs don't include the `parser` option (the
    // Prettier docs recommend against setting it and letting Prettier infer
    // based on the file-name), we include `...defaultPrettierOptions` here
    // before `...prettierConfig`.
    return prettier.format(template, {
      ...defaultPrettierOptions,
      ...prettierConfig,
    });
  } catch (err) {
    logger.warn(
      `Something went awry while trying to get the prettier config, falling back to default formatting [${err}]`
    );
    return prettier.format(template, defaultPrettierOptions);
  }
};
