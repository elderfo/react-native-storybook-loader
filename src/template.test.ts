import * as faker from 'faker';

import { generateTemplate } from './template';
import { LoaderDefinition } from './locator';

describe('generateTemplate', () => {
  test('generate story loader', async () => {
    const def: LoaderDefinition = {
      outputFile: faker.system.fileName(),
      storyFiles: [
        '../file1',
        '../sub/file2',
        '../../parent/file3',
        './sub/file4',
        './sub/sub/file5',
      ],
    };

    const contents = await generateTemplate(def);
    expect(contents).toMatchSnapshot();
  });
});
