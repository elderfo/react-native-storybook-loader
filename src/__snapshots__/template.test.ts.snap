// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTemplate generate story loader 1`] = `
"// Auto-generated file created by react-native-storybook-loader
// Do not edit.
//
// https://github.com/elderfo/react-native-storybook-loader.git

function loadStories() {
  require('../file1');
  require('../sub/file2');
  require('../../parent/file3');
  require('./sub/file4');
  require('./sub/sub/file5');
}

const stories = [
  '../file1',
  '../sub/file2',
  '../../parent/file3',
  './sub/file4',
  './sub/sub/file5',
];

const modules = [
  require('../file1'),
  require('../sub/file2'),
  require('../../parent/file3'),
  require('./sub/file4'),
  require('./sub/sub/file5'),
];

module.exports = {
  loadStories,
  stories,
  modules,
};
"
`;
